File: rust/src/bin/top_ten_scorers.rs
use std::io::{self, BufRead};

struct Player {
    name: String,
    team: String,
    ppg: f64,
}

fn main() {
    let stdin = io::stdin();
    let players: Vec<Player> = stdin
        .lock()
        .lines()
        .filter_map(|line| line.ok())
        .filter_map(|line| {
            let fields: Vec<&str> = line.split(',').collect();
            let games: i32 = fields[2].parse().ok()?;
            let points: i32 = fields[3].parse().ok()?;
            if games < 15 {
                None
            } else {
                Some(Player {
                    name: fields[1].to_string(),
                    team: fields[0].to_string(),
                    ppg: points as f64 / games as f64,
                })
            }
        })
        .collect();

    let mut sorted_players = players;
    sorted_players.sort_by(|a, b| b.ppg.partial_cmp(&a.ppg).unwrap());

    for player in sorted_players.iter().take(10) {
        println!(
            "{:22}{:4}{:8.2}", player.name, player.team, player.ppg);
    }
}

File: clojure/top_ten_scorers.clj
(defrecord Player [team name games ppg])

(defn from-line [[team name games points]]
  (let [points (parse-double points)
        games (parse-long games)]
    (->Player team name games (/ points games))))

(defn player-report [player]
  (format "%-22s%-4s%8.2f" 
          (:name player) (:team player) (:ppg player)))

(dorun 
  (->> 
    (line-seq (java.io.BufferedReader. *in*))
    (map #(clojure.string/split % #","))
    (map from-line)
    (filter #(>= (:games %) 15))
    (sort-by :ppg >)
    (take 10)
    (map player-report)
    (map println)))
